File: src/mcp.ts
import { Behalf } from "./behalf.js";
import { Mandate } from "./mandate.js";
import { AuthorizationError } from "./errors.js";

/**
 * MCP middleware — the adoption killer. Wrap any MCP-style tool server once and
 * every tool call is automatically checked against the caller's mandate: scope,
 * limits, expiry, revocation, and audit, with no per-tool code.
 */

/** Minimal structural view of an MCP-ish tool server. */
export interface ToolServerLike {
  callTool(name: string, args: Record<string, unknown>, ctx?: CallContext): Promise<unknown>;
}

/** Per-call context; the caller's mandate rides here. */
export interface CallContext {
  mandate?: Mandate;
  [k: string]: unknown;
}

/** A capability requirement: a fixed string or one derived from call args. */
export type CapabilityRule = string | ((args: Record<string, unknown>) => string);

export interface WithBehalfOptions {
  /** Map each tool name to the capability it requires. */
  policy: Record<string, CapabilityRule>;
  /** What to do on a denied call. */
  onDenied?: "throw" | "prompt";
  /**
   * Just-in-time consent callback (used when onDenied === "prompt"). Return true
   * to allow the call through despite the missing authority.
   */
  onPrompt?: (info: { tool: string; capability: string; mandate?: Mandate }) => Promise<boolean>;
  /** Engine to use (defaults to the shared Behalf default). */
  engine?: Behalf;
}

/**
 * Wrap a tool server so every call is authorized first.
 *
 *   const server = withBehalf(myMcpServer, {
 *     policy: {
 *       send_email: "write:email",
 *       read_calendar: "read:calendar",
 *       transfer_funds: (args) => `spend:usd<=${args.amount}`,
 *     },
 *     onDenied: "throw",
 *   });
 */
export function withBehalf(server: ToolServerLike, opts: WithBehalfOptions): ToolServerLike {
  return {
    async callTool(name, args, ctx) {
      const rule = opts.policy[name];
      // Tools without a policy entry pass through unguarded (explicit opt-in).
      if (rule === undefined) return server.callTool(name, args, ctx);

      const capability = typeof rule === "function" ? rule(args) : rule;
      const mandate = ctx?.mandate;

      let denied: string | undefined;
      if (!mandate) {
        denied = "no mandate presented";
      } else {
        try {
          await mandate.authorize(capability);
        } catch (e) {
          denied = e instanceof AuthorizationError ? e.reason : String(e);
        }
      }

      if (denied) {
        if (opts.onDenied === "prompt" && opts.onPrompt) {
          const allow = await opts.onPrompt({ tool: name, capability, mandate });
          if (!allow) throw new AuthorizationError(capability, `${denied} (consent declined)`);
        } else {
          throw new AuthorizationError(capability, denied);
        }
      }

      return server.callTool(name, args, ctx);
    },
  };
}

/** Symmetric binding for agent-to-agent (A2A) calls — same enforcement shape. */
export function withBehalfA2A(server: ToolServerLike, opts: WithBehalfOptions): ToolServerLike {
  return withBehalf(server, opts);
}

/** A tool definition for the agent-legibility layer. */
export interface ToolDefinition {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  handler: (args: Record<string, unknown>, ctx?: CallContext) => Promise<unknown>;
}

/**
 * The three discovery tools that let any agent obtain and use authority
 * natively: `request_mandate`, `present_mandate`, and `check_authority`.
 */
export function behalfMcpTools(engine: Behalf = Behalf.default): ToolDefinition[] {
  return [
    {
      name: "request_mandate",
      description:
        "Request a scoped, time-bound mandate authorizing an agent to act on a principal's behalf.",
      inputSchema: {
        type: "object",
        required: ["principal", "agent", "can", "expiresIn"],
        properties: {
          principal: { type: "string" },
          agent: { type: "string" },
          can: { type: "array", items: { type: "string" } },
          expiresIn: { type: "string", description: 'e.g. "1h", "10m"' },
        },
      },
      handler: async (args) => {
        const m = engine.grant({
          principal: String(args.principal),
          agent: String(args.agent),
          can: args.can as string[],
          expiresIn: String(args.expiresIn),
        });
        return { mandate: m.serialize(), id: m.id };
      },
    },
    {
      name: "present_mandate",
      description: "Validate a serialized mandate and return its decoded scope, principal, and expiry.",
      inputSchema: {
        type: "object",
        required: ["mandate"],
        properties: { mandate: { type: "string" } },
      },
      handler: async (args) => {
        const m = engine.import(String(args.mandate));
        engine.verifySignature(m.token);
        return {
          valid: true,
          id: m.id,
          principal: m.principal,
          agent: m.agent,
          expiresAt: m.expiresAt,
          chain: m.chain,
        };
      },
    },
    {
      name: "check_authority",
      description:
        "Advisory check of whether a mandate's scope would allow an action (does not perform it and does not prove possession).",
      inputSchema: {
        type: "object",
        required: ["mandate", "action"],
        properties: { mandate: { type: "string" }, action: { type: "string" } },
      },
      handler: async (args) => {
        const m = engine.import(String(args.mandate));
        return engine.inspect(m.token, String(args.action));
      },
    },
  ];
}

File: python/behalf/mcp.py
"""MCP / A2A middleware — wrap a tool server once; every call is authorized."""

from __future__ import annotations

from typing import Awaitable, Callable, Optional, Union

from .behalf import Behalf
from .errors import AuthorizationError
from .mandate import Mandate

# A capability rule: a fixed string or a callable deriving one from call args.
CapabilityRule = Union[str, Callable[[dict], str]]
ToolHandler = Callable[..., object]


def with_behalf(
    server,
    *,
    policy: dict[str, CapabilityRule],
    on_denied: str = "throw",
    on_prompt: Optional[Callable[[dict], bool]] = None,
):
    """Wrap a tool server so every call is authorized first.

    ``server`` must expose ``call_tool(name, args, ctx=None)``. The caller's
    mandate rides on ``ctx["mandate"]``. Tools absent from ``policy`` pass
    through unguarded (explicit opt-in).
    """

    class _Guarded:
        def call_tool(self, name: str, args: dict, ctx: Optional[dict] = None):
            rule = policy.get(name, _MISSING)
            if rule is _MISSING:
                return server.call_tool(name, args, ctx)

            capability = rule(args) if callable(rule) else rule
            mandate: Optional[Mandate] = (ctx or {}).get("mandate")

            denied: Optional[str] = None
            if mandate is None:
                denied = "no mandate presented"
            else:
                try:
                    mandate.authorize(capability)
                except AuthorizationError as e:
                    denied = e.reason

            if denied:
                if on_denied == "prompt" and on_prompt is not None:
                    allow = on_prompt(
                        {"tool": name, "capability": capability, "mandate": mandate}
                    )
                    if not allow:
                        raise AuthorizationError(capability, f"{denied} (consent declined)")
                else:
                    raise AuthorizationError(capability, denied)

            return server.call_tool(name, args, ctx)

    return _Guarded()


# Symmetric binding for agent-to-agent calls — same enforcement shape.
with_behalf_a2a = with_behalf

_MISSING = object()


def behalf_mcp_tools(engine: Optional[Behalf] = None) -> list[dict]:
    """The three discovery tools: request_mandate, present_mandate, check_authority."""
    eng = engine or Behalf.default()

    def request_mandate(args: dict) -> dict:
        m = eng.grant(
            principal=args["principal"],
            agent=args["agent"],
            can=args["can"],
            expires_in=args["expiresIn"],
        )
        return {"mandate": m.serialize(), "id": m.id}

    def present_mandate(args: dict) -> dict:
        m = eng.import_(args["mandate"])
        eng.verify_signature(m.token)
        return {
            "valid": True,
            "id": m.id,
            "principal": m.principal,
            "agent": m.agent,
            "expiresAt": m.expires_at,
            "chain": m.chain,
        }

    def check_authority(args: dict) -> dict:
        m = eng.import_(args["mandate"])
        try:
            m.authorize(args["action"])
            return {"allowed": True}
        except AuthorizationError as e:
            return {"allowed": False, "reason": e.reason}

    return [
        {
            "name": "request_mandate",
            "description": "Request a scoped, time-bound mandate authorizing an agent.",
            "inputSchema": {
                "type": "object",
                "required": ["principal", "agent", "can", "expiresIn"],
                "properties": {
                    "principal": {"type": "string"},
                    "agent": {"type": "string"},
                    "can": {"type": "array", "items": {"type": "string"}},
                    "expiresIn": {"type": "string"},
                },
            },
            "handler": request_mandate,
        },
        {
            "name": "present_mandate",
            "description": "Validate a serialized mandate and return its decoded scope.",
            "inputSchema": {
                "type": "object",
                "required": ["mandate"],
                "properties": {"mandate": {"type": "string"}},
            },
            "handler": present_mandate,
        },
        {
            "name": "check_authority",
            "description": "Check whether a mandate authorizes an action (does not perform it).",
            "inputSchema": {
                "type": "object",
                "required": ["mandate", "action"],
                "properties": {"mandate": {"type": "string"}, "action": {"type": "string"}},
            },
            "handler": check_authority,
        },
    ]
